[quantization] Throw if someone tries to torch.save() quantized modules

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26828 [quantization] Throw if someone tries to torch.save() quantized modules**

Pickle serialization for quantized modules is currently broken by https://github.com/pytorch/pytorch/issues/24045, so let's be loud and fail if the user tries to do it

Differential Revision: [D17579127](https://our.internmc.facebook.com/intern/diff/D17579127)

Commit:
[quantization] Throw if someone tries to torch.save() quantized modules

driazati commented:
Everything that gets `torch.save`d gets funnelled through the `persistent_id` function in `serialization.py`, you can do it there and avoid all of this sadness

```diff
diff --git a/torch/serialization.py b/torch/serialization.py
index 348602df05..c80b40a393 100644
--- a/torch/serialization.py
+++ b/torch/serialization.py
@@ -303,6 +303,10 @@ def _save(obj, zip_file, pickle_module, pickle_protocol):
         # see
         # https://docs.python.org/2/library/pickle.html#pickling-and-unpickling-external-objects
         # https://github.com/python/cpython/blob/master/Lib/pickle.py#L527-L537
+        import inspect
+        if inspect.isclass(obj) and issubclass(obj, torch.nn.Module):
+            if hasattr(obj, '__getstate__') or hasattr(obj, '__setstate__'):
+                print("bad!")
         if torch.is_storage(obj):
             storage_type = normalize_storage_type(type(obj))
             # Offset is always 0, but we keep it for backwards compatibility
```

jamesr66a commented:
@driazati putting this in the classes itself is meant to limit the blast radius of the change. We don't want to accidentally break someone's (otherwise working) code with a change like that

jamesr66a commented:
@driazati for a concrete example, your proposal breaks the following:

```
import torch

class ModuleNoTensors(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.some_int = 3

    def forward(self, x):
        return x + self.some_int

    def __getstate__(self):
        return (self.some_int,)

    def __setstate__(self, state):
        self.some_int = state[0]

m = ModuleNoTensors()

torch.save(m, 'foo.pkl')
loaded = torch.load('foo.pkl')

print('some int', loaded.some_int)
```

Which works as of today

driazati commented:
You could only throw for quantized models with something like `torch.nn.quantized in inspect.getmodule(obj).__name__`

jamesr66a commented:
@driazati I don't see the point of baking this into the framework itself if it's only for two specific modules, then

zdevito reviewed (approved):
This code quality is top notch.

jamesr66a commented:
Welp, this triggers ubsan so looks like i'm gonna have to change the approach

Commit:
Update on "[quantization] Throw if someone tries to torch.save() quantized modules"


Pickle serialization for quantized modules is currently broken by https://github.com/pytorch/pytorch/issues/24045, so let's be loud and fail if the user tries to do it

Differential Revision: [D17579127](https://our.internmc.facebook.com/intern/diff/D17579127)

jamesr66a commented:
@driazati showed me `torch.jit.is_scripting()` so now all the hackyness woes are gone. @dzhulgakov are you happy now
